### ISSUE TITLE
More indexing options

### ISSUE THREAD

**Issue by jl-wynen** [scipp/scipp#1815]
It would be useful to allow full slices without specifying a dimension. For instance to simplify assignments like
```.py
a['x', 1:2].coords['x'][:] = 1 * sc.units.one
```
There is technically no need to require a dimension label in the third pair of brackets.

Additionally, we should support the Ellipsis:
```.py
v['x', ...] = 1 * sc.units.one
```

**Comment by SimonHeybrock** [comment]
For the first example, why not use
```python
a.coords['x']['x', 1:2] = 1 * sc.units.one
```
?

**Comment by jl-wynen** [comment]
Maybe the example isn't great but that case would still benefit from the addition.

Alternatively:
```.py
v[:] = other
```
the reason for allowing omission of the dimension is that this way, we could set all elements of `v` without having to look up its dimensions manually while still getting unit and variance handling as opposed to 
```.py
v.values = other.values
```

**Comment by SimonHeybrock** [comment]
I suppose this would basically remove the need for exposing the `copy(src, dst)` function to Python?

**Comment by SimonHeybrock** [comment]
> I suppose this would basically remove the need for exposing the `copy(src, dst)` function to Python?

Scratch that, `copy` is different, since it also copies coords.

**Comment by SimonHeybrock** [comment]
Now that #1915 is merged, the `[:]` syntax becomes considerably more relevant in practice:

```python
da.data = var  # replace data variable
da.data[:] = var  # assign to slice, copy into existing data variable
```

**Comment by nvaytet** [comment]
It would also be nice to be able to use smart indexing, e.g. things like
```
selection = var['x', [1, 5, 7]]
```
which is super useful on `numpy`.

**Comment by SimonHeybrock** [comment]
> It would also be nice to be able to use smart indexing, e.g. things like
> 
> ```
> selection = var['x', [1, 5, 7]]
> ```
> 
> which is super useful on `numpy`.

What would that return?

**Comment by jl-wynen** [comment]
It should be the same as numpy: https://numpy.org/doc/stable/reference/arrays.indexing.html#advanced-indexing

I just realised that numpy's advanced indexing essentially has a way for applying masks which would translate into
```.py
da['x', ~da.masks['x']]
```
assuming that `da.masks['x']` is only in dim `'x'`. Maybe this is useful?

**Comment by SimonHeybrock** [comment]
Is "the same as numpy" returning a copy? I suppose this is more or less what you can do using `groupby` (but without a coord)?

**Comment by SimonHeybrock** [comment]
> I just realised that numpy's advanced indexing essentially has a way for applying masks which would translate into
> 
> ```python
> da['x', ~da.masks['x']]
> ```
> 
> assuming that `da.masks['x']` is only in dim `'x'`. Maybe this is useful?

This is just `groupby`, I think.

**Comment by jl-wynen** [comment]
> Is "the same as numpy" returning a copy?

Yes. Otherwise you would need a very sophisticated view which is not supported by the buffer protocol.

> This is just groupby, I think.

Nothing in this issue is fundamentally new or impossible in scipp now (I think), this is more about providing a familiar syntax.

**Comment by SimonHeybrock** [comment]
>     * [ ]  Without label
>       It would be useful to allow full slices without specifying a dimension. For instance to simplify assignments like
>       ```python
>       a['x', 1:2].coords['x'][:] = 1 * sc.units.one
>       ```      
>       There is technically no need to require a dimension label in the third pair of brackets.

As discussed previously:
- Can be done using `[...]`, so simplification is not a reason any more.
- `[:]` is ambiguous and scipp *always* requires dimension labels for readability.

I suggest to remove this item from the list.

**Comment by jl-wynen** [comment]
Fine by me

**Comment by g5t** [comment]
Just to expand on the multi-dimensional indexing idea. One _can_ achieve the same effect via a function like,
```python
def at(array, dims, subscripts):
    for dim, idx in zip(dims, subscripts):
        array = array[dim, idx]
    return array
```
which allows for usage,
```python
subarray = at(array, dims, subscripts)
first_scalar_element = at(array, array.dims, [0 for _ in array.shape])
```
but I can only imagine that this is not an optimal implementation, and in any case it would be nice to have the syntax
```python
subarray = array[dims, subscripts]
```

**Comment by SimonHeybrock** [comment]
I think we would probably use a dict syntax for indexing in multiple dimensions at once, just like `xarray`:

```python
da1 = array[dict(a=slice(1,2),b=5, c=6)]
da2 = array[dict(zip(dims, subscripts))]
```

**Comment by SimonHeybrock** [comment]
Useful examples:
```python
da = da['pixel', ~da.masks['pixel']]
```

**Comment by SimonHeybrock** [comment]
Multi-dim not done yet, open new issue if required.
